internal/controller/delayed_restart_manager.go

```diff
@@ -0,0 +1,291 @@
+	ErrNilRestartFunc = errors.New("restartFunc cannot be nil")
+)
+
+// TargetKey uniquely identifies a restart target (deployment/statefulset)
+type TargetKey struct {
+	Kind      string
+	Name      string
+	Namespace string
+}
+
+// String returns a string representation of the TargetKey for logging
+func (t TargetKey) String() string {
+	return fmt.Sprintf("%s/%s/%s", t.Kind, t.Namespace, t.Name)
+}
+
+// RestartFunc is a function that performs the actual restart operation
+type RestartFunc func(ctx context.Context) error
+
+// DelayedRestartManager manages delayed restart operations for workloads
+type DelayedRestartManager interface {
+	// ScheduleRestart schedules a delayed restart for a target
+	// Returns the final delay being used (in case it's already scheduled with a different delay)
+	// and whether this is a new schedule (false if already scheduled)
+	ScheduleRestart(ctx context.Context, target TargetKey, delay time.Duration, restartFunc RestartFunc) (finalDelay time.Duration, isNew bool, err error)
+
+	// IsScheduled checks if a target already has a pending delayed restart
+	IsScheduled(target TargetKey) bool
+
+	// Cancel cancels a scheduled restart for a target
+	Cancel(target TargetKey)
+
+	// Shutdown gracefully shuts down the manager, canceling all pending restarts
+	Shutdown()
+}
+
+// delayedTask represents a scheduled restart task
+type delayedTask struct {
+	target      TargetKey
+	delay       time.Duration
+	restartFunc RestartFunc
+	timer       *time.Timer
+	cancelCtx   context.Context
+	cancelFunc  context.CancelFunc
+	scheduledAt time.Time
+	executeAt   time.Time
+}
+
+// delayedRestartManager is the concrete implementation of DelayedRestartManager
+type delayedRestartManager struct {
+	mu           sync.RWMutex
+	tasks        map[TargetKey]*delayedTask
+	shutdownOnce sync.Once
+	logger       logr.Logger
+}
+
+// NewDelayedRestartManager creates a new DelayedRestartManager instance
+//
+//nolint:ireturn // Factory function returning interface is intentional
+func NewDelayedRestartManager(logger logr.Logger) DelayedRestartManager {
+	return &delayedRestartManager{
+		tasks:  make(map[TargetKey]*delayedTask),
+		logger: logger,
+	}
+}
+
+// ScheduleRestart schedules a delayed restart for a target
+func (m *delayedRestartManager) ScheduleRestart(ctx context.Context, target TargetKey, delay time.Duration, restartFunc RestartFunc) (finalDelay time.Duration, isNew bool, err error) {
+	if delay < 0 {
+		return 0, false, fmt.Errorf("%w: %v", ErrNegativeDelay, delay)
+	}
+
+	if restartFunc == nil {
+		return 0, false, ErrNilRestartFunc
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Check if there's already a scheduled restart for this target
+	existingTask, exists := m.tasks[target]
+	if exists {
+		// Target already has a pending restart
+		if delay > existingTask.delay {
+			// New delay is higher, reschedule with the higher delay
+			m.logger.Info("Rescheduling restart with higher delay",
```

Comment:
Contrary to issue: "If a change triggers a restart of an already delayed target, the target should not be restarted but only an info message should be logged"